substrate/frame/sassafras/src/lib.rs

```diff
@@ -0,0 +1,1059 @@
+	/// The configuration for the next epoch.
+	#[pallet::storage]
+	#[pallet::getter(fn next_config)]
+	pub type NextEpochConfig<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Pending epoch configuration change that will be set as `NextEpochConfig` when the next
+	/// epoch is enacted.
+	///
+	/// In other words, a config change submitted during epoch N will be enacted on epoch N+2.
+	/// This is to maintain coherence for already submitted tickets for epoch N+1 that where
+	/// computed using configuration parameters stored for epoch N+1.
+	#[pallet::storage]
+	pub type PendingEpochConfigChange<T> = StorageValue<_, EpochConfiguration>;
+
+	/// Stored tickets metadata.
+	#[pallet::storage]
+	pub type TicketsMeta<T> = StorageValue<_, TicketsMetadata, ValueQuery>;
+
+	/// Tickets identifiers map.
+	///
+	/// The map holds tickets ids for the current and next epoch.
+	///
+	/// The key is a tuple composed by:
+	/// - `u8` equal to epoch's index modulo 2;
+	/// - `u32` equal to the ticket's index in a lexicographically sorted list of epoch's tickets.
+	///
+	/// Epoch X first N-th ticket has key (X mod 2, N)
+	///
+	/// Note that the ticket's index doesn't directly correspond to the slot index within the epoch.
+	/// The assigment is computed dynamically using an *outside-in* strategy.
+	///
+	/// Be aware that entries within this map are never removed, only overwritten.
+	/// Last element index should be fetched from the [`TicketsMeta`] value.
+	#[pallet::storage]
+	pub type TicketsIds<T> = StorageMap<_, Identity, (u8, u32), TicketId>;
+
+	/// Tickets to be used for current and next epoch.
+	#[pallet::storage]
+	pub type TicketsData<T> = StorageMap<_, Identity, TicketId, TicketBody>;
+
+	/// Next epoch tickets unsorted segments.
+	///
+	/// Contains lists of tickets where each list represents a batch of tickets
+	/// received via the `submit_tickets` extrinsic.
+	///
+	/// Each segment has max length [`SEGMENT_MAX_SIZE`].
+	#[pallet::storage]
+	pub type UnsortedSegments<T: Config> =
+		StorageMap<_, Identity, u32, BoundedVec<TicketId, ConstU32<SEGMENT_MAX_SIZE>>, ValueQuery>;
+
+	/// The most recently set of tickets which are candidates to become the next
+	/// epoch tickets.
+	#[pallet::storage]
+	pub type SortedCandidates<T> =
+		StorageValue<_, BoundedVec<TicketId, MaxTicketsFor<T>>, ValueQuery>;
+
+	/// Parameters used to construct the epoch's ring verifier.
+	///
+	/// In practice: Updatable Universal Reference String and the seed.
+	#[pallet::storage]
+	#[pallet::getter(fn ring_context)]
+	pub type RingContext<T: Config> = StorageValue<_, vrf::RingContext>;
+
+	/// Ring verifier data for the current epoch.
+	#[pallet::storage]
+	pub type RingVerifierData<T: Config> = StorageValue<_, vrf::RingVerifierData>;
+
+	/// Slot claim vrf-preoutput used to generate per-slot randomness.
+	///
+	/// The value is ephemeral and is cleared on block finalization.
+	#[pallet::storage]
+	pub(crate) type ClaimTemporaryData<T> = StorageValue<_, vrf::VrfOutput>;
```

Comment:
I guess this can be used to implement `ParentBlockRandomness` as in BABE. For that case we'd need to keep this in the state instead of clearing it in `on_finalize`. Just a thought, nothing to do in this PR right now.

Reply:
https://github.com/paritytech/polkadot-sdk/pull/1577#issuecomment-1812707909